common/src/crypto.rs

```diff
@@ -32,20 +44,45 @@ impl fmt::Display for Hash256 {
 /// A cryptographic signature.
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize)]
 pub struct Signature {
-    pub dummy: String,
+    signature: Vec<u8>,
 }
 
 impl Signature {
     /// Creates a new signature from the given data and keys.
-    pub fn sign(_data: Hash256, _public_key: &PublicKey, _private_key: &PrivateKey) -> Self {
-        Signature {
-            dummy: "".to_string(),
-        }
+    pub fn sign(
+        data: Hash256,
+        public_key: &PublicKey,
+        private_key: &PrivateKey,
+    ) -> Result<Self, Error> {
+        let public_key = ed25519_dalek::PublicKey::from_bytes(&public_key.key)
+            .map_err(|_| Error::InvalidFormat(format!("public key: {}", public_key)))?;
+        let private_key = ed25519_dalek::SecretKey::from_bytes(&private_key.key)
+            .map_err(|_| Error::InvalidFormat("private key: [omitted]".to_owned()))?;
+        let signature = ed25519_dalek::Keypair {
+            secret: private_key,
+            public: public_key,
+        };
+        let signature = signature.sign(data.hash.as_ref());
+        Ok(Signature {
+            signature: signature.to_bytes().to_vec(),
+        })
     }
 
     /// Verifies the signature against the given data and public key.
-    pub fn verify(&self, _data: Hash256, _public_key: &PublicKey) -> bool {
-        true
+    pub fn verify(&self, data: Hash256, public_key: &PublicKey) -> Result<(), Error> {
```

Comment:
Why different format compared to `fn sign`?

Reply:
It seems we need both pubkey and privkey (a keypair) to sign a message using ed25519_dalek.
See: https://docs.rs/ed25519-dalek/1.0.1/ed25519_dalek/trait.Signer.html

Reply:
@lemonholic I meant the line changes in the function prototype for `fn sign`

Reply:
I don't get it. Do you mean that the change of using `Result`?

Reply:
No, I was just talking about bunch of line changes in `fn sign` while not in other functions

Reply:
Insertion of new lines is automatically done by rustfmt.